File: iosApp/iosApp/offer/validation/EditedSettlementMethodValidator.swift
//
//  NewSettlementMethodValidator.swift
//  iosApp
//
//  Created by jimmyt on 8/6/22.
//  Copyright © 2022 orgName. All rights reserved.
//

/**
 Validates edited settlement methods for an offer made by the user of this interface.
 
 This function ensures that:
    - the user has selected at least one settlement method
    - the user has specified a price for each selected settlement method
    - the user has supplied their information for each selected settlement method.
 
 - Parameter settlementMethods: The `Array` of `SettlementMethod`s to be validated.
 
 - Returns: An `Array` of validated `SettlementMethod`s
 
 - Throws An `EditedSettlementMethodValidationError` if this is not able to ensure any of the conditions in the list above. The descriptions of the errors thrown by this function are human-readable and can be displayed to the user so that they can correct any problems.
 */
func validateEditedSettlementMethods(
    _ settlementMethods: [SettlementMethod]
) throws -> [SettlementMethod] {
    guard !settlementMethods.isEmpty else {
        throw EditedSettlementMethodValidationError(desc: "You must specify at leat one settlement method.")
    }
    try settlementMethods.forEach { settlementMethod in
        guard settlementMethod.price != "" else {
            throw EditedSettlementMethodValidationError(desc: "You must specify a price for each settlement method you select.")
        }
        guard settlementMethod.privateData != nil else {
            throw NewOfferDataValidationError(desc: "You must supply your information for each settlement method you select.")
        }
    }
    return settlementMethods
}

File: androidApp/src/main/java/com/commuto/interfacemobile/android/offer/validation/EditedSettlementMethodValidator.kt
package com.commuto.interfacemobile.android.offer.validation

import com.commuto.interfacemobile.android.settlement.SettlementMethod

/**
 * Validates edited settlement methods for an offer made by the user of this interface.
 *
 * This function ensures that:
 *  - the user has selected at least one settlement method
 *  - the user has specified a price for each selected settlement method
 *
 *  @param settlementMethods The [List] of [SettlementMethod]s to be validated.
 *
 *  @return A [List] of validated [SettlementMethod]s.
 *
 *  @throws [EditedSettlementMethodValidationException] if this is not able to ensure any of the conditions in the list
 *  above. The messages of the exceptions thrown by this function are human-readable and can be displayed to the user so
 *  that they can correct any problems.
 */
fun validateEditedSettlementMethods(
    settlementMethods: List<SettlementMethod>
): List<SettlementMethod> {
    if (settlementMethods.isEmpty()) {
        throw EditedSettlementMethodValidationException("You must specify at least one settlement method.")
    }
    settlementMethods.forEach {
        if (it.price == "") {
            throw EditedSettlementMethodValidationException("You must specify a price for each selected settlement " +
                    "method.")
        }
    }
    return settlementMethods
}